staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go

```diff
@@ -182,6 +182,12 @@ func UpdateResource(r rest.Updater, scope *RequestScope, admit admission.Interfa
 			})
 		}
 
+		// Ignore changes that only affect managed fields
+		// timestamps. FieldManager can't know about changes
+		// like normalized fields, defaulted fields and other
+		// mutations.
+		transformers = append(transformers, fieldmanager.IgnoreManagedFieldsTimestampsTransformer)
+
```

Comment:
I share @liggitt 's concerns that, while code-wise being clean now, this fixup logic still feels misplaced. Why does SSA even update the managed fields timestamp? Why do we need this fixup code being put in every update/patch code patch, even without SSA being involved?

To understand, this is the order:
1. SSA does its thing
2. mutating admission runs
3. defaulting runs
4. the resulting object migth be unchanged compared to oldObj, while step (1) saw a change?

Reply:
In other words, this transformer fights against an SSA patch that is continuesly trying to fight against some admission plugin or default. E.g. a mutating admission plugin sets `spec.privileged=false` for security policy reasons, but a GitOps system insists on `spec.privileged=true`. The result without this PR would be a hot loop with constantly updated managed fields timestamp.

Reply:
One maybe important detail for a questionable gut feeling: 

This logic uses semantic equality, which is uncommon elsewhere in the storage stack. What if we used normal, deep equality?

Then this would turn into:

a mutating admission plugin sets `spec.extraPermissions=[]` for security policy reasons, but a GitOps system insists on `spec.extraPermissions=undefined`. The result without this PR would be a hot loop with constantly updated managed fields timestamp.

Reply:
On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook set undefined to `[]`) where the managedField is not updated.

Reply:
> Why does SSA even update the managed fields timestamp (my take below)? Why do we need this fixup code being put in every update/patch code patch, even without SSA being involved?

Because we don't want to track defaults/mutation changes as part of SSA so we run at the very beginning, but it also means that we don't know know what these changes are going to be and if they will undo a change.

> To understand, this is the order:
> 1. SSA does its thing
> 2. mutating admission runs
> 3. defaulting runs
> 4. the resulting object migth be unchanged compared to oldObj, while step (1) saw a change?

Yes, that's correct. Step 1 doesn't know what step 2 and 3 will do and specifically doesn't want to track these changes. But specifically when 2 and 3 undo what 1 saw, then we want to skip the extra etcd write (hence 4).

> The result without this PR would be a hot loop with constantly updated managed fields timestamp.

That's correct, that's the problem we're trying to solve here, and it's causing actual pain to users.

> This logic uses semantic equality, which is uncommon elsewhere in the storage stack. What if we used normal, deep equality

Note that this comparison is not by-passing any write to etcd. The only action it takes now is to by-pass the update to the timestamp in the managed fields.

We can not use deep equality because deep equality doesn't work for structures that have unexported fields.
In order to make it work, we had to copy the whole semantic deepequal (which doesn't do anything but provide custom comparison for kubernetes types that have unexported fields: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go#L29-L49). It also doesn't do anything specific for `undefined to []`.

> On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook sets undefined to []) where the managedField is not updated.

No, because we're not changing the object or preventing the object from going to etcd, we're just reverting the timestamp change. So if there is a change in the object, it's still going to go to etcd. If there is a change to managed fields other than the timestamp, it's still going to go to etcd.

Reply:
> It also doesn't do anything specific for undefined to [].
```
// An empty slice *is* equal to a nil slice for our purposes; same for maps.
```

> On the other hand, with this PR in place we would miss actual changes in etcd (where a mutating webhook sets undefined to []) where the managedField is not updated.

I meant that the object changes in etcd, but `managedFields` is not set. "miss" in that sense.

Reply:
The latest revision includes a new function `Equalities.DeepEqualWithoutImplicitEqualities` which should not allow implicit equality for nil/empty slices/maps. I believe that resolves concerns about the empty slices/maps.